src/crepe/system/CollisionSystem.cpp

```diff
@@ -1,5 +1,393 @@
+#include <cmath>
+#include <algorithm>
+#include <cstddef>
+#include <functional>
+#include <utility>
+#include <variant>
+#include <optional>
+
+#include "api/Event.h"
+#include "api/EventManager.h"
+#include "api/BoxCollider.h"
+#include "api/CircleCollider.h"
+#include "api/Vector2.h"
+#include "api/Rigidbody.h"
+#include "api/Transform.h"
+
+#include "ComponentManager.h"
 #include "CollisionSystem.h"
+#include "Collider.h"
+#include "types.h"
+#include "util/OptionalRef.h"
 
 using namespace crepe;
 
-void CollisionSystem::update() {}
+void CollisionSystem::update() {
+	// Get collider components and keep them seperate
+	ComponentManager & mgr = this->component_manager;
+	std::vector<std::reference_wrapper<BoxCollider>> boxcolliders	= mgr.get_components_by_type<BoxCollider>();
+	std::vector<std::reference_wrapper<CircleCollider>> circlecolliders	= mgr.get_components_by_type<CircleCollider>();
+
+	std::vector<collider_variant> all_colliders;
+	// Add BoxCollider references
+	for (auto& box : boxcolliders) {
+			all_colliders.push_back(collider_variant{box});
+	}
+
+	// Add CircleCollider references
+	for (auto& circle : circlecolliders) {
+			all_colliders.push_back(collider_variant{circle});
+	}
+	
+	// Check between all colliders if there is a collision
+	std::vector<std::pair<CollisionInternal,CollisionInternal>> collided = check_collisions(all_colliders);
+
+	// For both objects call the collision handler 
+	for (auto& collision_pair : collided) {
+		collision_handler_request(collision_pair.first,collision_pair.second);
+		collision_handler_request(collision_pair.second,collision_pair.first);
+	}
+}
+
+void CollisionSystem::collision_handler_request(CollisionInternal& data1,CollisionInternal& data2){
+
+	CollisionInternalType type = check_collider_type(data1.collider,data2.collider);
+	std::pair<vec2,CollisionSystem::Direction> resolution_data = collision_handler(data1,data2,type);
+
+	OptionalRef<Collider> collider1;
+	OptionalRef<Collider> collider2;
+	switch (type) {
```

Comment:
Ik ben geen grote fan van hoe je de verschillende kruisingen hier weer omzet naar de echte typen, maar volgensmij plannen we niet om andere collidervormen te ondersteunen dan cirkel en rechthoek dus dan is dit denk ik beter om zo te laten staan want lui.

Reply:
Wat is trouwens het nut van alle `std::get` met de concrete types als je ze vervolgens weer opslaat in een reference naar de basisklasse?

Reply:
Dit is puur gedaan zodat de user in een script bij de collider kan als die dat wilt. dus in de oncollision in het script krijg je een struct mee van data die je kunt gebruiken. en omdat ik de collider al heb opgevraagd hoeft de user dit niet te doen.